shell/platform/darwin/macos/framework/Source/FlutterRenderer.mm

```diff
@@ -95,30 +95,34 @@ - (FlutterMetalTexture)createTextureForView:(uint64_t)viewId size:(CGSize)size {
     // FlutterMetalTexture has texture `null`, therefore is discarded.
     return FlutterMetalTexture{};
   }
-  FlutterRenderBackingStore* backingStore = [view backingStoreForSize:size];
-  id<MTLTexture> texture = backingStore.texture;
-  FlutterMetalTexture embedderTexture;
-  embedderTexture.struct_size = sizeof(FlutterMetalTexture);
-  embedderTexture.texture = (__bridge void*)texture;
-  embedderTexture.texture_id = reinterpret_cast<int64_t>(texture);
-  return embedderTexture;
+  FlutterSurface* surface = [view.surfaceManager surfaceForSize:size];
+  FlutterMetalTexture texture = surface.asFlutterMetalTexture;
+  // This is here because the non-compositor path completely ignores
+  // destruction and callback. It is ugly, but the surface manager
+  // will keep the borrowed surface retained until present: is called.
+  if (texture.destruction_callback != nullptr) {
+    texture.destruction_callback(texture.user_data);
+  }
```

Comment:
The root problem here is that the non compositor path of embedder throws away the `user_data` and `destruction_callback` fields in `FlutterMetalTexture`. If we could fix that and ensure that when `FlutterMetalPresentCallback` is called, the user_data is intact, we could remove the "borrowed surfaces" bookkeeping from surface manager completely.

Reply:
In [this commit](https://github.com/flutter/engine/pull/37789/commits/27f0891bf7f571f0143b88cbb28561744e2f283f), I made sure that non compositing path properly remember user data and calls the destruction callback. I'm not sure that's the best implementation, but it needs to be done. I'm not too happy with that shared_ptr there. I tried to make `GPUMTLTextureInfo` non copyable, but unfortunately it is captured by `std::function`, which is copyable so this doesn't work.

Still, this needs to be fixed one way or another, otherwise the user_data in passed to the present callback is just garbage, which is really bad.

With that in place, I could remove any bookkeeping of borrowed surfaces. So the `FlutterSurface` looks like this now:

```objc
@interface FlutterSurface : NSObject

- (FlutterMetalTexture)asFlutterMetalTexture;
+ (nullable FlutterSurface*)fromFlutterMetalTexture:(nonnull const FlutterMetalTexture*)texture;

@end
```

And the general flow is
```objc

FlutterMetalTexture GetTexture() {
   FlutterSurface *surface = [surfaceManager surfaceForSize: size];
   return [surface asFlutterMetalTexture];
}

void PresentMetalTexture(const FlutterMetalTexture *texture) {
   FlutterSurface *surface = [FlutterSurface fromFlutterMetalTexture:texture];
  [surfaceManager present: ...];
}
```

Reply:
To update this thread, the embedder API release bug was tracked in this bug and is now fixed: https://github.com/flutter/flutter/issues/116381